embassy-rp/src/clocks.rs

```diff
@@ -15,124 +172,411 @@ pub(crate) unsafe fn init() {
+            w.set_auxsrc(conf.src);
+        });
+    } else {
+        peris.set_adc(false);
+    }
+
+    if let Some(conf) = config.rtc_clk {
+        // CLK RTC = PLL USB (48MHz) / 1024 = 46875Hz
+        c.clk_rtc_ctrl().modify(|w| {
+            w.set_enable(false);
+        });
+        c.clk_rtc_div().write(|w| {
+            w.set_int(conf.div_int);
+            w.set_frac(conf.div_frac);
+        });
+        c.clk_rtc_ctrl().write(|w| {
+            w.set_enable(true);
+            w.set_auxsrc(conf.src);
+        });
+    } else {
+        peris.set_rtc(false);
+    }
+
+    // Peripheral clocks should now all be running
+    reset::unreset_wait(peris);
+}
+
+unsafe fn configure_rosc(config: RoscConfig) {
+    let p = pac::ROSC;
+
+    p.freqa().write(|w| {
+        w.set_passwd(pac::rosc::vals::Passwd::PASS);
+        w.set_ds0(config.drive_strength[0]);
+        w.set_ds1(config.drive_strength[1]);
+        w.set_ds2(config.drive_strength[2]);
+        w.set_ds3(config.drive_strength[3]);
     });
 
-    // CLK RTC = PLL USB (48MHz) / 1024 = 46875Hz
-    c.clk_rtc_ctrl().modify(|w| {
-        w.set_enable(false);
+    p.freqb().write(|w| {
+        w.set_passwd(pac::rosc::vals::Passwd::PASS);
+        w.set_ds4(config.drive_strength[4]);
+        w.set_ds5(config.drive_strength[5]);
+        w.set_ds6(config.drive_strength[6]);
+        w.set_ds7(config.drive_strength[7]);
     });
-    c.clk_rtc_div().write(|w| w.set_int(1024));
-    c.clk_rtc_ctrl().write(|w| {
-        w.set_enable(true);
-        w.set_auxsrc(ClkRtcCtrlAuxsrc::CLKSRC_PLL_USB);
+
+    p.div().write(|w| {
+        w.set_div(pac::rosc::vals::Div(config.div + pac::rosc::vals::Div::PASS.0));
     });
 
-    // CLK PERI = clk_sys. Used as reference clock for Peripherals. No dividers so just select and enable
-    // Normally choose clk_sys or clk_usb
-    c.clk_peri_ctrl().write(|w| {
-        w.set_enable(true);
-        w.set_auxsrc(ClkPeriCtrlAuxsrc::CLK_SYS);
+    p.ctrl().write(|w| {
+        w.set_enable(pac::rosc::vals::Enable::ENABLE);
+        w.set_freq_range(config.range);
     });
+}
 
-    // Peripheral clocks should now all be running
-    let peris = reset::ALL_PERIPHERALS;
-    reset::unreset_wait(peris);
+pub fn estimate_rosc_freq() -> u32 {
+    let p = pac::ROSC;
+
+    let base = match unsafe { p.ctrl().read().freq_range() } {
+        pac::rosc::vals::FreqRange::LOW => 84_000_000,
+        pac::rosc::vals::FreqRange::MEDIUM => 104_000_000,
+        pac::rosc::vals::FreqRange::HIGH => 140_000_000,
+        pac::rosc::vals::FreqRange::TOOHIGH => 208_000_000,
+        _ => unreachable!(),
+    };
+    let mut div = unsafe { p.div().read().0 - pac::rosc::vals::Div::PASS.0 as u32 };
+    if div == 0 {
+        div = 32
+    }
+
+    base / div
 }
 
-pub(crate) fn _clk_sys_freq() -> u32 {
-    125_000_000
+pub fn xosc_freq() -> u32 {
+    unsafe { XIN_HZ }
 }
 
-pub(crate) fn clk_peri_freq() -> u32 {
-    125_000_000
+pub fn gpin0_freq() -> u32 {
```

Comment:
if these are TODO i'd just remove them

Reply:
I'd like to leave this in. The alternative is that in the various functions that get clock frequencies we either todo for those two sources, or they just sorta vanish into the `_ => unreachable!()`, which is worse.

Long term this gets messy, as only the application knows what the gpin frequency is. Some kind of shared state is probably necessary.

Reply:
ah okay! I had missed that it's used by us, not just the user.